### ISSUE TITLE
False positive in rust-analyzer diagnostics when macro produces a variable that might shadow

### ISSUE THREAD

**Issue by iitalics** [rust-lang/rust-analyzer#19054]
rust-analyzer diagnostics seems to become confused about macro hygiene in some edge cases, causing false positives. if we "shadow" a local variable such that it changes the mutability of a reference type (`&mut T` -> `&T`) then rust-analyzer thinks this is an issue. however, this isn't really shadowing because macro hygiene should have protected the inner variable from being exposed the caller. if we completely change the type while shadowing then rust-analyzer correctly does not report an error. in each of these cases rustc correctly does not report an error.

**rust-analyzer version**: (eg. output of "rust-analyzer: Show RA Version" command, accessible in VSCode via <kbd>Ctrl/⌘</kbd>+<kbd>Shift</kbd>+<kbd>P</kbd>)

rust-analyzer 1.80.0 (0514789 2024-07-21)

**rustc version**: (eg. output of `rustc -V`)

rustc 1.80.0 (051478957 2024-07-21)

**editor or extension**: (eg. VSCode, Vim, Emacs, etc. For VSCode users, specify your extension version; for users of other editors, provide the distribution if applicable)

Emacs

**relevant settings**: (eg. client settings, or environment variables like `CARGO`, `RUSTC`, `RUSTUP_HOME` or `CARGO_HOME`)

**repository link (if public, optional)**: (eg. [rust-analyzer](https://github.com/rust-lang/rust-analyzer))

**code snippet to reproduce**: 
```rust
macro_rules! hygienic {
    () => {
        // this variable is local to the macro, so it should not actually shadow definitions
        // in caller-scope
        let hygiene: &i32 = &0;
        let _ = hygiene;
    };
}

fn main() {
    // this block has no issues with rustc or rust-analyzer
    {
        let whatever = &mut 0;
        hygienic!();
        let _: &mut i32 = whatever;
    }

    // this block is identical to above, modulo variable renaming
    {
        let hygiene = &mut 0;
        hygienic!();
        // rust-analyzer produces the following error:
        //
        //     expected &mut i32, found &i32
        //
        // but no error is produced by rustc or cargo check
        let _: &mut i32 = hygiene;
    }

    // this block is similar to above but we changed the local variable to have type
    // bool. it has no issues
    {
        let hygiene = true;
        hygienic!();
        let _: bool = hygiene;
    }
}
```

**Comment by ChayimFriedman2** [comment]
I can't reproduce this (VSCode). However hygiene was somewhat recently implemented, so it may not be in rustup 1.80.0.
